### PR TITLE
My own version of Gnuboy. New functionalities added

### PR DESCRIPTION
**Pull request by ghost**
-Screenshots
-Screencasting
-Console/debugger (beta)

A console will appear on the screen when Gnuboy is running. Type 'help' and you will see the available commands and how to take both screenshots and do screencasts.

Important: enable debugging in order to use a few of the functionalities within the debugger. Otherwise, they won't work (Error messages will appear, while normal execution will go on without any results).

### CODE REVIEW

**Comment by rofl0r** [comment]
cool!
i'm generally willing to add your enhancements, but merging a single 1000 lines patch is against all my principles. therefore i'd like to ask you to go over your changes and make one commit per new feature or logical change, so we can bring each commit to a size that's reasonable to review.
does that sound acceptable?

File: cpucore.h [hunk 1]
```diff
@@ -7,7 +7,7 @@ const static byte cycles_table[256] =
 	1, 3, 2, 2, 1, 1, 2, 1, 5, 2, 2, 2, 1, 1, 2, 1,
 	1, 3, 2, 2, 1, 1, 2, 1, 3, 2, 2, 2, 1, 1, 2, 1,
 	3, 3, 2, 2, 1, 1, 2, 1, 3, 2, 2, 2, 1, 1, 2, 1,
-	3, 3, 2, 2, 3, 3, 3, 1, 3, 2, 2, 2, 1, 1, 2, 1,
+	3, 3, 2, 2, 1, 3, 3, 3, 3, 2, 2, 2, 1, 1, 2, 1,
```
Comment (rofl0r) [inline cpucore.h:10]: why was this changed? please put into a single commit explaining.

File: Makefile.in [hunk 2]
```diff
@@ -26,7 +26,7 @@ FB_LIBS =
 SVGA_OBJS = sys/svga/svgalib.o sys/pc/keymap.o @JOY@ @SOUND@
```
Comment (rofl0r) [inline Makefile.in:26]: please no gratuitous modechanges, thanks.
(refering to  Makefile.in 100644 → 100755, and the same for almost all other files)

File: README [hunk 3]
```diff
@@ -45,20 +45,19 @@ all of the following and much more:
   Wave pattern memory corruption when sound channel 3 is played.
   Pad, timer, divide counter, and other basic hardware registers.
   CGB double-speed CPU mode.
-  Sorting sprites by X coordinate in DMG mode.
```
Comment (rofl0r) [inline README:48]: why was this block moved to section below?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# mygnuboy
+My own version of gnuboy emulator
```
Comment (rofl0r) [inline README.md:2]: i dont think we want to change the name to "mygnuboy". it should stay as close to laguna's original as possible.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        -pthread -----> I have implemented pthreads, so it is necessary too
+
+
+NEW DEPENDENCIES ADDED:
+
+
+	scrot
+	ffmpeg
+
+	Both of them available via apt-get
```
Comment (rofl0r) [inline README.md:22]: not all the world is debian/ubuntu.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+	-fforce-mem  -----> must be removed if compilation is not successful
+	-std=c99 -----> necessary because there are a few comments starting with //
+        -pthread -----> I have implemented pthreads, so it is necessary too
+
+
+NEW DEPENDENCIES ADDED:
+
+
+	scrot
+	ffmpeg
```
Comment (rofl0r) [inline README.md:20]: if you want to add third-party dependency, then please make them optional. ffmpeg is a huge download and takes considerable time to build.

File: cpu.c [hunk 7]
```diff
@@ -641,7 +643,7 @@ int cpu_emulate(int cycles)
 		break;
 	case 0x3C: /* INC A */
 		INC(A); break;
-			
+
```
Comment (rofl0r) [inline cpu.c:646]: please no gratuitous whitespace changes.

File: cpu.h [hunk 8]
```diff
@@ -26,16 +25,10 @@ struct cpu
...
 extern struct cpu cpu;
 
-void cpu_reset();
-void div_advance(int cnt);
-void timer_advance(int cnt);
-void lcdc_advance(int cnt);
-void sound_advance(int cnt);
-void cpu_timers(int cnt);
-int cpu_emulate(int cycles);
+int return_globalpc();
```
Comment (rofl0r) [inline cpu.h:31]: what's the point of removing these decls ?
it looks more and more as if you just copied the "community edition" gnuboy over this one.

File: debian/changelog [hunk 9]
```diff
@@ -0,0 +1,113 @@
+gnuboy (1.0.3-7) unstable; urgency=low
+
+  * Acknowledging previous NMUs, thanks Moritz! closes: #711112. 
+
+ -- Davide Puricelli (evo) <evo@debian.org>  Sun, 30 Jun 2013 16:54:28 +0200
+
```
Comment (rofl0r) [inline debian/changelog:6]: i dont think debian specific files have any reason to go into this repo.

**Review by rofl0r — CHANGES_REQUESTED** [review]
i added some comments. i'll do further review once you've come up with clean one-feature-per-commit commits.

**Comment by rofl0r** [comment]
since you deleted your fork of my repo, i assume you're not interested of cleaning up your work. thanks for nothing.

**Comment by ghost** [comment]
Hello rofl0r!

Of course I am interested on it, but since this is the first time I ever participate in a free software project and I realised I did a few things wrong, I will work on what I wanted to commit so that it fits and respects all the things you remarked before. 

I still want to contribute with a debugger functionality. Regarding both the screencast and the snapshot functionalities, which could be an appropriate approach when trying to keep the program available for several distributions?

I will try to commit every change so that everyone can keep track of what I have done in a proper way.

Thank you, and sorry for the inconvenience :)

**Comment by rofl0r** [comment]
oh ok. did you see the other issue #1 ? 
basically there are 2 versions of gnuboy: the one that i call in my mind the "mob" version, which is what debian has in their repos. it is the community's continuation of laguna's project, which has some new features and changes. then there's this version which is basically the pure version of when laguna abandoned the project, plus some minor fixes to make it work on current systems. afaict, your work is based on the community/mob version; so maybe that one would be better suited to file a PR against.

**Comment by rofl0r** [comment]
> Regarding both the screencast and the snapshot functionalities, which could be an appropriate approach when trying to keep the program available for several distributions?

oh sorry, i overlooked the question in my previous reply.
what other (well-behaving) programs do in such a situation is to assume user wants all features, and then test if the required dependencies for each feature are available; if not, turn it off.
for example in case of feature "record video": a configure option --enable-ffmpeg which defaults to "auto" which means a compile test is done against ffmpeg.
this could either involve using `pkg-config` to check for its existence, or using `gcc helloworld.c -lavcodec`. if the latter succeeds it signifies that ffmpeg is installed; but then you also need to check for its headers, that would mean compiling a test program which includes a ffmpeg header such as:
```c
#include <libavcodec/avcodec.h>
int main() { return 0; }
```
if the user passes --disable-ffmpeg, than all tests are skipped and ffmpeg support is not compiled in.
in the code this would look like
```c
#ifdef WANT_FFMPEG
record_video();
#endif
```

**Comment by rofl0r** [comment]
btw regarding the snapshot functionality:
there's already a screen buffer that has the entire bitmap that's put on the screen. what prevents you from just saving that to disk directly (for example using a .bmp header), instead of pulling in a dependency on scrot ?
